Restrict media privacy until a referencing page is published (#46)
ONSdigital/dis-wagtail: docs/custom-features/media_privacy.md

# Media privacy

This project uses customised versions of Wagtail's `Image` and `Document` models to store and manage the
privacy of uploaded media files.

1. The privacy of an image or document cannot be changed manually via the Wagtail UI.
2. Images and documents are private by default, and will only be served to authenticated users with some level of permission on the media item whilst they have this privacy status (so previewing from Wagtail should still work as expected).
3. Where the storage backend supports it, files for private images and documents (and any renditions created from the original image) will have their access permissions updated at the storage level, preventing direct file access.
4. Images and documents only become public when they are referenced by blocks in page content, and that page is published.
5. When a page is unpublished, any public images or documents referenced solely by that page are made private again (references from other draft pages are ignored, as the media can easily be made public again when the draft page is published).

## Use of Wagtail's reference index

> Wagtail's reference index is used to track when images and documents (and other objects) are referenced by pages and other objects.

Reviewer: suggestion:

```suggestion
Wagtail's [reference index](https://docs.wagtail.org/en/stable/advanced_topics/reference_index.html) is used to track when images and documents (and other objects) are referenced by pages and other objects.
```

Author: Applied in [04e31c2](https://github.com/ONSdigital/dis-wagtail/pull/46/commits/04e31c2223dbc6004a1397f19ddaadee04101b93)

> Population of the reference index is automatically triggered by the `post_save` signal for all registered models.

Reviewer: how about

```suggestion
The reference index is automatically populated via the `post_save` signal for all registered models.
```

?

Author: Applied with slight wording change in [04e31c2](https://github.com/ONSdigital/dis-wagtail/pull/46/commits/04e31c2223dbc6004a1397f19ddaadee04101b93)

When publishing or unpublishing a page (or other type of 'publishable' object), object-level changes are saved before any `published`, `page_published`, `unpublished` or `page_unpublished` signals are emitted. So, as long as we use these signals to review the privacy
of referenced media, we can be sure that the reference index will be up-to-date.

## Handling of file permission setting errors

Whilst file-permission setting requests are quite reliable in S3, they can fail occasionally (as with any web service).

Failed attempts are tracked using a `file_permissions_last_set` timestamp for each media item, which is only updated if all update requests for the media item were successful.

> The `retry_file_permission_set_attempts` management command seeks out any media items with an outdated timestamp and attempts to update the files again.

Reviewer: Note to self: we will need a page in docs that lists the management commands and how often they run. One for a follow up PR

of referenced media, we can be sure that the reference index will be up-to-date.

## Handling of file permission setting errors

Whilst file-permission setting requests are quite reliable in S3, they can fail occasionally (as with any web service).

Failed attempts are tracked using a `file_permissions_last_set` timestamp for each media item, which is only updated if all update requests for the media item were successful.

The `retry_file_permission_set_attempts` management command seeks out any media items with an outdated timestamp and attempts to update the files again.

The custom `PrivacySettingS3Storage` class logs failed file-setting attempts using the 'WARNING' log level, so should be available to review in most environments should issues occur.

## Compatibility with upstream caching

Because serve view responses and direct file urls can be cached, it's important that when the privacy of a media item changes, the cache
is invalidated for any affected URLs. In environments where Wagtail's front-end-cache app is configured, all relevant URLs should be collected into a batch and sent to the relevant backend to initiate purging.

> NOTE: Image serve view responses are not cached, so don't need to be invalidated.

Reviewer 2: Issue: This seems to be incorrect: https://github.com/wagtail/wagtail/blob/3e75c018be642eae97eeba22051041280f285825/wagtail/images/views/serve.py#L38

Author: I believe that is cancelled-out by the fact that the url is included in `private_urls` in `urls.py`, which applies `never_cache` to each item

Reviewer 2: Might be worth adding a test to make sure this is the case.

Author: Now we're caching image urls again, this point is redundant. @RealOrangeOne are you happy to mark this as resolved?

# Media privacy

This project uses customised versions of Wagtail's `Image` and `Document` models to store and manage the
privacy of uploaded media files, so that potentially sensitive media isn't leaked to the public before it's ready.

## The basic premise

1. Images and documents are 'private' by default, and will only be served to authenticated users with some level of permission on the media item.
2. They only become 'public' when a page referencing them is published.
3. When a live page unpublished, any images or documents referenced soley by that page (i.e. not used on any other live pages) are made private again.

When a media item is 'private', file requests are handled by views that check the permissions of the requesting user, and only serve the file if the user has some level of permission on the media item (to support previews and reviews from other editors).

When a media item is 'public', permission-checking in serve views is skipped, and the user is either redirected to the media storage URL (in the case of images), or served the file directly (in the case of documents).

> These cases (and more) are covered by integration tests in [cms/private_media/tests/test_signal_handlers.py](https://github.com/ONSdigital/dis-wagtail/tree/main/cms/private_media/tests/test_signal_handlers.py).

Author: Not yet, but it will once merged :)
